pylops/waveeqprocessing/_kirchhoff_cuda.py

```diff
@@ -0,0 +1,519 @@
+       utilizing GPU acceleration through Numba's CUDA capabilities.
+
+       Parameters
+       ----------
+       ns : int
+           Number of sources.
+       nr : int
+           Number of receivers.
+       nt : int
+           Number of time samples.
+       ni : int
+           Number of image points.
+       dynamic : int, optional
+           Flag indicating whether to use dynamic computation. ``True`` == 1 or not ``False`` == 0 (default is 0).
+       travsrcrec : int, optional
+           Flag indicating whether to use separate tables for src and rec traveltimes. Seperate == 1   (default is 0).
+       """
+
+    def __init__(self, ns, nr, nt, ni, dynamic=0, travsrcrec=0):
+        self.dynamic, self.travsrcrec = dynamic, travsrcrec
+        self.ns, self.nr, self.nt, self.ni = ns, nr, nt, ni
+
+        self._lunch_grid_setup()
+
+    def _lunch_grid_setup(self):
+        """ Set up the CUDA grid and block dimensions based on the current device and computation flags.
+                This method configures the number of blocks and threads per block for CUDA kernels,
+                depending on whether dynamic computation and travel times from sources and receivers are used.
+        """
+        current_device = cuda.get_current_device()
+        if self.dynamic:
+            num_sources = self.ns
+            num_streams = 3
+            streams = [cuda.stream() for _ in range(num_streams)]
+            sources_per_stream = num_sources // num_streams
+            remainder = num_sources % num_streams
+            source_counter = 0
+            self.sources_per_streams = {}
+            for i, stream in enumerate(streams):
+                num_sources_for_stream = sources_per_stream + (1 if i < remainder else 0)
+                sources_for_stream = list(range(source_counter, source_counter + num_sources_for_stream))
+                self.sources_per_streams[stream] = sources_for_stream
+                source_counter += num_sources_for_stream
+            n_runs = num_streams * self.ns * self.ni * self.nr  # number_of_times_to_run_kernel
+            self.num_threads_per_blocks = current_device.WARP_SIZE * 8
+            self.num_blocks = (n_runs + (self.num_threads_per_blocks - 1)) // self.num_threads_per_blocks
+            # print(num_threads_per_blocks)
+            # print(self.blocks)
+        else:
+            if not self.travsrcrec:
+                # version 4
+                self.num_threads_per_blocks = current_device.WARP_SIZE * 8
+                self.num_blocks = ((self.ns * self.nr) + (
+                            self.num_threads_per_blocks - 1)) // self.num_threads_per_blocks
+            else:
+                # version 3
+                wrap = current_device.WARP_SIZE
+                multipr_count = current_device.MULTIPROCESSOR_COUNT
+                self.num_threads_per_blocks = (wrap, wrap)
+                self.num_blocks = (multipr_count, multipr_count)
```

Comment:
I am unsure about this. For my GPU (RTX 3090), the output of this:
```
from numba import cuda

current_device = cuda.get_current_device()
current_device.WARP_SIZE, current_device.MULTIPROCESSOR_COUNT
```
gives (32,82). This means that indipendently on the number of sources and receivers, you always have a 82x82 grid with 32x32 threads, which in many cases will call many blocks that will never pass the `if ns > isrc and nr > irec:` condition. Am I getting this wrong?

Reply:
In this case, each block always has a fixed number of threads, specifically `(32, 32)`. The number of blocks, however, depends on the total number of threads required. This implementation ensures that no block is allocated unless it will run at least one thread.

However, I need to fix the last else as it does not follow the same approach ..

Reply:
Not sure I understand your reply. My point is that you code will ALWAYS allocate the same number of blocks no matter the size of the problem.

Take this code:
```
 wrap = current_device.WARP_SIZE
multipr_count = current_device.MULTIPROCESSOR_COUNT
self.num_threads_per_blocks = (wrap, wrap)
self.num_blocks = (multipr_count, multipr_count)
```
on my GPU (RTX 3090), I get `self.num_threads_per_blocks = (32, 32)` and `self.num_blocks = (82, 82)`. Now later, you do:
```
self._trav_kirch_matvec_cuda[self.num_blocks, self.num_threads_per_blocks](x_d, y_d, nsnr_d, nt_d, ni_d, itrav_d, travd_d)
```
this will ALWAYS call 82x82 blocks no matter how many sources or receivers you have in your problem. So say for 64 sources and 64 receivers, you will have the first 2x2 blocks doing something and all the other running but just never passing this condition `if ns > isrc and nr > irec:`.

In the new code, i changed this to:
```
self.num_blocks = ((self.ns + self.num_threads_per_blocks[0] - 1
                                ) // self.num_threads_per_blocks[0], 
                                (self.nr + self.num_threads_per_blocks[1] - 1
                                ) // self.num_threads_per_blocks[1])
```
which is actually similar to what you had done for case 4. Here we don't launch any block that will not contribute to the final solution

Reply:
Thank you for correcting the kernel launch, as I did not follow the same approach as in the other cases.

I missed this part and ended up copying a lot of code because I didn't fork the repo before starting to make changes.